internal/relation/endpoint.go

```diff
@@ -0,0 +1,47 @@
+// Copyright 2025 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package relation
+
+import (
+	"fmt"
+
+	"github.com/juju/juju/internal/charm"
+)
+
+// CounterpartRole returns the RelationRole that this RelationRole
+// can relate to.
+// This should remain an internal method because the relation
+// model does not guarantee that for every role there will
+// necessarily exist a single counterpart role that is sensible
+// for basing algorithms upon.
+func CounterpartRole(r charm.RelationRole) charm.RelationRole {
+	switch r {
+	case charm.RoleProvider:
+		return charm.RoleRequirer
+	case charm.RoleRequirer:
+		return charm.RoleProvider
+	case charm.RolePeer:
+		return charm.RolePeer
+	}
+	panic(fmt.Errorf("unknown relation role %q", r))
+}
+
+// Endpoint represents one endpoint of a relation.
+type Endpoint struct {
+	ApplicationName string
+	charm.Relation
```

Comment:
Would it be better to have this struct explicitly in a field rather than embedded? It would help avoid confusion when reading other bits of the code as to what is a property of the relation definition vs a property of its instantiation.

Reply:
Perhaps, however at this point, it'd require many code changes, especially in state code which will be deleted in a few weeks.  If we want, I'd postpone that work until after state has been deleted.